File: python/src/shardloom/runtime_defaults.py
"""Shared local runtime resource defaults for ShardLoom Python surfaces."""

from __future__ import annotations

import os


SHARDLOOM_MEMORY_GB_ENV = "SHARDLOOM_MEMORY_GB"
SHARDLOOM_MAX_PARALLELISM_ENV = "SHARDLOOM_MAX_PARALLELISM"


def _positive_int_env(name: str, default: int, *, floor: int = 1) -> int:
    raw = os.environ.get(name)
    if raw is None:
        return default
    try:
        value = int(raw)
    except ValueError:
        return default
    return max(value, floor) if value > 0 else default


DEFAULT_LOCAL_RUNTIME_MEMORY_GB = _positive_int_env(SHARDLOOM_MEMORY_GB_ENV, 4)
DEFAULT_LOCAL_RUNTIME_MAX_PARALLELISM = _positive_int_env(
    SHARDLOOM_MAX_PARALLELISM_ENV,
    2,
    floor=2,
)
DEFAULT_INTERNAL_SMOKE_MEMORY_GB = 1
DEFAULT_INTERNAL_SMOKE_MAX_PARALLELISM = 1

File: shardloom-cli/src/runtime_defaults.rs
//! Shared runtime resource defaults for public local `ShardLoom` surfaces.

pub(crate) const DEFAULT_PUBLIC_LOCAL_RUNTIME_MEMORY_GB: u64 = 4;
pub(crate) const DEFAULT_PUBLIC_LOCAL_RUNTIME_MAX_PARALLELISM: usize = 2;
pub(crate) const MIN_PUBLIC_LOCAL_RUNTIME_MAX_PARALLELISM: usize = 2;

pub(crate) const PUBLIC_LOCAL_RUNTIME_MEMORY_GB_ENV: &str = "SHARDLOOM_MEMORY_GB";
pub(crate) const PUBLIC_LOCAL_RUNTIME_MAX_PARALLELISM_ENV: &str = "SHARDLOOM_MAX_PARALLELISM";

pub(crate) fn default_public_local_runtime_memory_gb() -> u64 {
    std::env::var(PUBLIC_LOCAL_RUNTIME_MEMORY_GB_ENV)
        .ok()
        .and_then(|value| value.parse::<u64>().ok())
        .filter(|value| *value > 0)
        .unwrap_or(DEFAULT_PUBLIC_LOCAL_RUNTIME_MEMORY_GB)
}

pub(crate) fn default_public_local_runtime_max_parallelism() -> usize {
    std::env::var(PUBLIC_LOCAL_RUNTIME_MAX_PARALLELISM_ENV)
        .ok()
        .and_then(|value| value.parse::<usize>().ok())
        .filter(|value| *value > 0)
        .unwrap_or(DEFAULT_PUBLIC_LOCAL_RUNTIME_MAX_PARALLELISM)
        .max(MIN_PUBLIC_LOCAL_RUNTIME_MAX_PARALLELISM)
}
